// @orca-so/common-sdk/dist/web3/transactions/transactions-builder.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilder = void 0;
const web3_js_1 = require("@solana/web3.js");
const transactions_processor_1 = require("./transactions-processor");
/**
 * @category Transactions Util
 */
class TransactionBuilder {
    constructor(provider) {
        this.provider = provider;
        this.instructions = [];
        this.signers = [];
    }
    addInstruction(instruction) {
        this.instructions.push(instruction);
        return this;
    }
    addSigner(signer) {
        this.signers.push(signer);
        return this;
    }
    isEmpty() {
        return this.instructions.length == 0;
    }
    /**
     * Compresses all instructions & signers in this builder
     * into one single instruction
     * @param compressPost Compress all post instructions into the instructions field
     * @returns Instruction object containing all
     */
    compressIx(compressPost) {
        let instructions = [];
        let cleanupInstructions = [];
        let signers = [];
        this.instructions.forEach((curr) => {
            instructions = instructions.concat(curr.instructions);
            // Cleanup instructions should execute in reverse order
            cleanupInstructions = curr.cleanupInstructions.concat(cleanupInstructions);
            signers = signers.concat(curr.signers);
        });
        if (compressPost) {
            instructions = instructions.concat(cleanupInstructions);
            cleanupInstructions = [];
        }
        return {
            instructions: [...instructions],
            cleanupInstructions: [...cleanupInstructions],
            signers,
        };
    }
    /**
     * Constructs a transaction payload with the gathered instructions
     * @returns a TransactionPayload object that can be excuted or agregated into other transactions
     */
    // TODO: Anchor 0.24+ removes .wallet from Provider
    build(options = { latestBlockhash: false }) {
        return __awaiter(this, void 0, void 0, function* () {
            const { latestBlockhash } = options;
            let recentBlockhash;
            if (latestBlockhash === true) {
                recentBlockhash = (yield this.provider.connection.getLatestBlockhash("singleGossip"))
                    .blockhash;
            }
            else if (latestBlockhash !== false && latestBlockhash.blockhash) {
                recentBlockhash = latestBlockhash.blockhash;
            }
            const transaction = new web3_js_1.Transaction({
                recentBlockhash,
                feePayer: this.provider.wallet.publicKey,
            });
            const ix = this.compressIx(true);
            transaction.add(...ix.instructions);
            transaction.feePayer = this.provider.wallet.publicKey;
            return {
                transaction: transaction,
                signers: ix.signers.concat(this.signers),
            };
        });
    }
    /**
     * Constructs a transaction payload with the gathered instructions, sign it with the provider and send it out
     * @returns the txId of the transaction
     */
    buildAndExecute() {
        return __awaiter(this, void 0, void 0, function* () {
            const tx = yield this.build();
            const tp = new transactions_processor_1.TransactionProcessor(this.provider);
            const { execute } = yield tp.signAndConstructTransaction(tx);
            return execute();
        });
    }
}
exports.TransactionBuilder = TransactionBuilder;
